flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/FileSystemTableSink.java

```diff
@@ -373,16 +374,16 @@ public DynamicTableSource.DataStructureConverter createDataStructureConverter(
         };
     }
 
-    private Path toStagingPath() {
+    @VisibleForTesting
```

Comment:
You can increase the visibility of `toStagingPath` and add the `@VisibleForTesting` annotation.

Reply:
It's not the best solution. The classes are, as you already mentioned, hard to test and would require some bigger refactoring.

Reply:
Another workaround (actually not good either) might be
using `FileSystemTableSink`(and meanwhile introducing a test class extending it to return a constant path to reproduce the issue) to create the test harness in `FileSystemOutputFormatTest`. 

```java
    private Path toStagingPath() {
        // Add a random UUID to prevent multiple sinks from sharing the same staging dir.
        // Please see FLINK-29114 for more details
        Path stagingDir = new Path(path, ".staging_" + getStagingPathPostfix());
        try {
            FileSystem fs = stagingDir.getFileSystem();
            Preconditions.checkState(
                    fs.exists(stagingDir) || fs.mkdirs(stagingDir),
                    "Failed to create staging dir " + stagingDir);
            return stagingDir;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @VisibleForTesting
    protected String getStagingPathPostfix() {
        return System.currentTimeMillis() + "_" + UUID.randomUUID();
    }
```


```java
        private static class DummyFileSystemTableSink extends FileSystemTableSink {

        DummyFileSystemTableSink(
                ObjectIdentifier tableIdentifier,
                DataType physicalRowDataType,
                List<String> partitionKeys,
                ReadableConfig tableOptions,
                @Nullable DecodingFormat<BulkFormat<RowData, FileSourceSplit>> bulkReaderFormat,
                @Nullable DecodingFormat<DeserializationSchema<RowData>> deserializationFormat,
                @Nullable EncodingFormat<BulkWriter.Factory<RowData>> bulkWriterFormat,
                @Nullable EncodingFormat<SerializationSchema<RowData>> serializationFormat) {
            super(
                    tableIdentifier,
                    physicalRowDataType,
                    partitionKeys,
                    tableOptions,
                    bulkReaderFormat,
                    deserializationFormat,
                    bulkWriterFormat,
                    serializationFormat);
        }

        @Override
        protected String getStagingPathPostfix() {
            return "foo";
        }
    }
```
As you said before, we need a refactor to make things look not weird.